core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java

```diff
@@ -79,20 +137,100 @@ public void stop() throws MuleException {
+
+    private Scheduler beforeScheduler;
+    private Scheduler afterScheduler;
+
+    public SourcePolicyProcessingStrategy(SchedulerService schedulerService, SchedulerConfig schedulerBaseConfig,
+                                          Set<ProcessingType> beforeNext, Set<ProcessingType> afterNext,
+                                          Processor followingNext) {
+      this.schedulerService = schedulerService;
+      this.schedulerBaseConfig = schedulerBaseConfig;
+      this.beforeNext = beforeNext;
+      this.afterNext = afterNext;
+      this.followingNext = followingNext;
+    }
+
+    @Override
+    public void start() throws MuleException {
+      SchedulerConfig beforeConfig = schedulerBaseConfig.withName("policy-source-before");
+      if (beforeNext.stream().allMatch(p -> CPU_LITE_ASYNC.equals(p))) {
+        beforeScheduler = schedulerService.cpuLightScheduler(beforeConfig);
+      } else if (beforeNext.stream().anyMatch(p -> BLOCKING.equals(p) || IO_RW.equals(p))) {
+        beforeScheduler = schedulerService.ioScheduler(beforeConfig);
+      } else if (beforeNext.stream().anyMatch(p -> CPU_INTENSIVE.equals(p))) {
+        beforeScheduler = schedulerService.cpuIntensiveScheduler(beforeConfig);
+      }
+
+      SchedulerConfig afterConfig = schedulerBaseConfig.withName("policy-source-after");
+      if (afterNext.stream().allMatch(p -> CPU_LITE_ASYNC.equals(p))) {
+        afterScheduler = schedulerService.cpuLightScheduler(afterConfig);
+      } else if (afterNext.stream().anyMatch(p -> BLOCKING.equals(p) || IO_RW.equals(p))) {
+        afterScheduler = schedulerService.ioScheduler(afterConfig);
+      } else if (afterNext.stream().anyMatch(p -> CPU_INTENSIVE.equals(p))) {
+        afterScheduler = schedulerService.cpuIntensiveScheduler(afterConfig);
+      }
+    }
+
+    @Override
+    public void stop() throws MuleException {
+      if (beforeScheduler != null) {
+        beforeScheduler.stop();
+      }
+      if (afterScheduler != null) {
+        afterScheduler.stop();
+      }
+    }
+
+    @Override
+    public Sink createSink(FlowConstruct flowConstruct, ReactiveProcessor pipeline) {
+      return new StreamPerEventSink(pipeline, createOnEventConsumer());
+    }
+
+    @Override
+    public ReactiveProcessor onPipeline(ReactiveProcessor pipeline) {
+      return publisher -> Flux.from(publisher)
+          .publishOn(beforeScheduler != null ? fromExecutor(beforeScheduler) : immediate())
+          .transform(pipeline);
+    }
+
+
+    @Override
+    public ReactiveProcessor onProcessor(ReactiveProcessor processor) {
+      if (processor.equals(followingNext) || (processor instanceof InterceptedReactiveProcessor
+          && ((InterceptedReactiveProcessor) processor).getProcessor().equals(followingNext))) {
+        return publisher -> Flux.from(publisher)
+            .publishOn(afterScheduler != null ? fromExecutor(afterScheduler) : immediate())
+            .transform(processor);
+      } else {
```

Comment:
Need to handle the "after async_cpu_light" in onProcessor too, to ensure we don't do any policy processing in requester selector threads, if requesters are used in policies.